diff --git a/emscripten_browser_file.h b/emscripten_browser_file.h
index 7019883..45e6b16 100644
--- a/emscripten_browser_file.h
+++ b/emscripten_browser_file.h
@@ -37,7 +37,7 @@ EM_JS_INLINE(void, upload, (char const *accept_types, upload_handler callback, v
     file_reader.onload = (event) => {
       const uint8Arr = new Uint8Array(event.target.result);
       const data_ptr = Module["_malloc"](uint8Arr.length);
-      const data_on_heap = new Uint8Array(Module["HEAPU8"].buffer, data_ptr, uint8Arr.length);
+      const data_on_heap = new Uint8Array(HEAPU8.buffer, data_ptr, uint8Arr.length);
       data_on_heap.set(uint8Arr);
       Module["ccall"]('upload_file_return', 'number', ['string', 'string', 'number', 'number', 'number', 'number'], [event.target.filename, event.target.mime_type, data_on_heap.byteOffset, uint8Arr.length, callback, callback_data]);
       Module["_free"](data_ptr);
@@ -90,7 +90,10 @@ EM_JS_INLINE(void, download, (char const *filename, char const *mime_type, void
   /// Offer a buffer in memory as a file to download, specifying download filename and mime type
   var a = document.createElement('a');
   a.download = UTF8ToString(filename);
-  a.href = URL.createObjectURL(new Blob([new Uint8Array(Module["HEAPU8"].buffer, buffer, buffer_size)], {type: UTF8ToString(mime_type)}));
+  var bufferCopy = new ArrayBuffer(buffer_size);
+  var uint8Array = new Uint8Array(bufferCopy);
+  uint8Array.set(new Uint8Array(HEAPU8.buffer, buffer, buffer_size));
+  a.href = URL.createObjectURL(new Blob([uint8Array], {type: UTF8ToString(mime_type)}));
   a.click();
 });
 #pragma GCC diagnostic pop
